### PR TITLE
[Storage] Refactor badger cache and all the client code to use generics

### PR DESCRIPTION
**Pull request by nozim**
Re: https://github.com/onflow/flow-go/issues/3932

### CODE REVIEW

**Comment by Kay-Zee** [comment]
Thanks for the PR @nozim, I've gone ahead and kicked-off the CI for the branch, and will try to have someone look over the PR soon.

In the mean time, i had to update the branch with latest master to update a dependency that was causing problems for CI, and that seems to have broken tidy.

File: storage/badger/approvals.go
Comment (durkmurder) [inline storage/badger/approvals.go:41]:
```suggestion
			withStore(store),
			withRetrieve(retrieve)),
```

File: storage/badger/events.go
Comment (durkmurder) [inline storage/badger/events.go:159]:
```suggestion
			withRetrieve(retrieve)),
```

File: storage/badger/headers.go
Comment (durkmurder) [inline storage/badger/headers.go:62]:
```suggestion
			withStore(storeHeight),
			withRetrieve(retrieveHeight)),
```

File: storage/badger/transaction_results.go
Comment (durkmurder) [inline storage/badger/transaction_results.go:138]:
```suggestion
			withRetrieve(retrieve),
```

**Comment by durkmurder** [comment]
Looks good, did you try measuring performance impact of usings generics instead of `interface{}` ? We still haven't reached consensus on this topic inside the team. 

I would be interested to see at least synthetic benchmarks so we could compare old implementation vs new. Running localnet benchmark should also give some insights. Ideally I would like to see both.

**Comment by nozim** [comment]
@durkmurder good question, will do a small benchmark :)

**Comment by nozim** [comment]
Added small example comparing to interface implementation. @durkmurder fyi :)

**Comment by durkmurder** [comment]
@nozim Thank you, this is good to go. Can you remove those benchmarks and run lint and tidy?

FYI, you can try passing actual type instead of any and it will be even faster, it's very recognizable if you move `RandomBytes` outside the loop. 

I will approve this as soon after you cleanup it.
Good job!

**Comment by durkmurder** [comment]
@nozim just to make it clear, we don't want to merge the benchmark(and iface cache implementation) in master since we don't plan to use previous cache implementation.

**Comment by nozim** [comment]
@durkmurder yes, totally makes sense. Will clean it up as well ;)

File: storage/badger/qcs.go
Comment (peterargue) [inline storage/badger/qcs.go:26]: please remove if this isn't needed

File: storage/badger/generic_bench.txt
Comment (peterargue) [inline storage/badger/generic_bench.txt:1]: please remove this file too

**Review by peterargue — APPROVED** [review]
Thanks for cleaning this up @nozim! added a few small stylistic comments, but looks good.

**Comment by nozim** [comment]
@peterargue thanks for review, will address all of the notes )

**Review by peterargue — APPROVED** [review]
👍

**Comment by durkmurder** [comment]
@nozim can you fix last lint errors so I can approve? We can merge right after

**Review by durkmurder — APPROVED** [review]
:ship:

**Comment by nozim** [comment]
seems like a glitch in engine unit test?

**Pull request merged**
